Rename Rebuild button to Replay Now

felipecrs opened the pull request:
<!-- Please describe your pull request here. -->

May _close #618. Also related to #756.

I think the name Rebuild for this button was a poor chose.

That's because Rebuild is an existing concept of Jenkins, through the [Rebuilder Plugin](https://plugins.jenkins.io/rebuild/), and that's not at all what this button does.

Replay is another concept, and the regular build page already has a Replay button.

I think we should name this button as Replay to clearly reflect what it does and avoid any confusion/assumption.

**Before:**
![image](https://github.com/user-attachments/assets/da7ffa0c-2653-4105-8357-8044816cb350)


**After:**
![image](https://github.com/user-attachments/assets/e27ed726-dd9c-409f-ba0f-b6e8452e729f)


### Testing done

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

There's no much to test, it's working.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

timja commented:
I’m not sold on this, rebuild is what it does. 

Replay as a user facing term is really confusing. 

I get that it’s not a great situation for people who want to change parameters but it’s exactly the same experience as in core and in blue ocean.

yes a plugin was created which improves it but it’s the core experience which should be improved

felipecrs commented:
> Replay as a user facing term is really confusing.

But that's how the button is already named in Jenkins core:

![image](https://github.com/user-attachments/assets/75578f14-e76d-4835-9312-e62079bb8b71)

Blue Ocean invented a new name for it (_Re-run_), it never called it _Rebuild_.

And I disagree. They are not the same. When you click this button, the build ends up being replayed:

![image](https://github.com/user-attachments/assets/fcdf2d66-8a19-452e-9ad9-9d07d0481ac2)

There's the decades old _Build now_ or _Build with parameters_ button in the job page.

Intuitively, _Rebuild_ was supposed to behave similar to the _Build_ button, which is what the _Rebuilder plugin_ does.

And again, that's very different than Replay.

For example, when you Replay a build, it will inherit all environment variables injected dynamically by plugins like Gerrit Trigger.

When you Rebuild, it does not.

Each button provides a different functionality that should not be confused with each other.

felipecrs commented:
![image](https://github.com/user-attachments/assets/a5598feb-ff23-485b-8b8c-d962d613bc43)

<img width="1921" alt="chrome_c82s5fT7Nf" src="https://github.com/user-attachments/assets/5ba0866b-3076-4b15-8e66-cce5fec65788" />

timja commented:
Re-run or rebuild are really the same thing as a synonym though.

Think of it as a person who doesn't know about the rebuild plugin (as I did when I put that name there), What does replay mean? its not really an action as such imo.

To me Rebuild was clearest and seemed better than re-run for a label. Part of this is blue ocean also didn't add text just an icon for it.

@janfaracik have you got any thoughts?

felipecrs commented:
Blue Ocean has it through a tooltip:

![image](https://github.com/user-attachments/assets/175785a1-1b93-4f6c-951b-aa119da7a394)

To make it worse, even the icon of this plugin's _Rebuild_ button is the same as the one from _Rebuilder_ plugin

![image](https://github.com/user-attachments/assets/43a25cd9-33a2-40ae-aa71-9ee6bdb1c964)

![image](https://github.com/user-attachments/assets/94ebff70-a7ae-46d7-8f43-4999a655bf9c)

> Think of it as a person who doesn't know about the rebuild plugin (as I did when I put that name there)

That's completely understandable, but I hope you change your mind now that you are aware of the Rebuilder plugin. :)

> What does replay mean? its not really an action as such imo. To me Rebuild was clearest and seemed better than re-run for a label. 

I think this is an issue that should first be addressed in the Jenkins as a whole. We can discuss what's a better name for the Replay in Jenkins Core and change it there. We can discuss what should be done in the Rebuilder plugin too. And maybe even in Blue Ocean.

And while this issue is not addressed, I think the terminology should be fixed here. Who knows, maybe it will never be addressed.

timja commented:
FWIW Jenkins core doesn't have a replay concept, its the pipeline plugins that have it.

I take your points, need to discuss with @janfaracik.

felipecrs commented:
![image](https://github.com/user-attachments/assets/2f25b4d0-dbb8-40a9-84db-199407327960)

I changed the icon too.

stuartrowe commented:
I found the current `Rebuild` button confusing as well as a user of the [rebuild-plugin](https://github.com/jenkinsci/rebuild-plugin). 

I agree with @felipecrs that the pipeline graph view plugin should be consistent with the `replay` action contributed by the [workflow-cps-plugin](https://github.com/jenkinsci/workflow-cps-plugin) for pipelines. In my opinion it should actually redirect to the `<run-url>/replay` endpoint instead of just scheduling a re-run.

Ideally a secondary `rebuild` button could be enabled if its detected that the controller has the [rebuild-plugin](https://github.com/jenkinsci/rebuild-plugin) installed. This button would redirect to `<run-url>/rebuild` or <run-url>/rebuild/parameterized` depending on whether the build is parameterized. That only makes sense if the existing button is renamed first.

Unfortunately Jenkins Core doesn't provide a way to rebuild the current build immediately like the current `Rebuild` button does. Would it be terrible to overload what the button does?
- By default, use replay to reschedule the same build
- If the [rebuild-plugin](https://github.com/jenkinsci/rebuild-plugin) is installed make the button a redirect to `<run-url>/rebuild` or `<run-url>/rebuild/parameterized` depending on the job.

timja commented:
> In my opinion it should actually redirect to the <run-url>/replay

I really don't agree with this. As a user I just want the build to re-run, similar to in blue ocean.

I shouldn't have to go to a page seeing the pipeline.

lewisbirks commented:
> ![image](https://private-user-images.githubusercontent.com/29582865/443407430-2f25b4d0-dbb8-40a9-84db-199407327960.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcyMTIxNTUsIm5iZiI6MTc0NzIxMTg1NSwicGF0aCI6Ii8yOTU4Mjg2NS80NDM0MDc0MzAtMmYyNWI0ZDAtZGJiOC00MGE5LTg0ZGItMTk5NDA3MzI3OTYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTE0VDA4MzczNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwZjIyYWIzZjE3OTk4ZmU0ZGMxODYyMTY0OTQzODE2NDc1ZDQ2NThiMDQ2MTljMTlhYjM2NDgwZmRmMTM2NjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.m0bNa0rLjNfg2M1PrV1-L0gkHUBaWTKiF7RRHGdgCwY)
> 
> I changed the icon too.

If we want to change it to ensure that it isn't mistaken for anything but the replay functionality do we want to use the same icon as the [workflow-cps plugin](https://github.com/jenkinsci/workflow-cps-plugin/blob/7a8a8176d450b2cee72c68dba2335253f124f361/plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayAction.java#L109)?

felipecrs commented:
Yeah. You're right. Too bad that icon is too ugly. Maybe we can change the workflow plugin to use a newer icon? lol

felipecrs commented:
I think I addressed all concerns now:

![image](https://github.com/user-attachments/assets/e59926b1-8d4d-436a-a762-5c60c9706648)

Same icon, and renamed to _Replay Now_, similar to:

![image](https://github.com/user-attachments/assets/6863da7e-4a9c-492b-817f-9e9b19e62b0c)

felipecrs commented:
> Would it be terrible to overload what the button does?
> 
> * By default, use replay to reschedule the same build
> * If the [rebuild-plugin](https://github.com/jenkinsci/rebuild-plugin?rgh-link-date=2025-05-14T07%3A22%3A26Z) is installed make the button a redirect to `<run-url>/rebuild` or `<run-url>/rebuild/parameterized` depending on the job.

I think it would be terrible, yes. I personally use Replay and Rebuild interchangeably, given they do different things. Instead, I suggest:

- https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/756

felipecrs commented:
> * Rebuild

What will _Rebuild_ do? If it does _Replay_, then the fundamental problem this PR is based on is still unsolved.

And I'm confused about the second button _Build Now_. Are we still talking about the _Pipeline Overview_ page?

lhupfeldt commented:
I'm with @felipecrs. Replay and Rebuild is not the same. Please name it `Replay` if sticking with a single button.
A dropdown choice with more explanation would be great.

felipecrs commented:
@mawinter69 what you are suggesting is exactly what I suggested here:

- https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/756

What remains to be resolved for this PR is whether we rename the current button to reflect the reality or not, while this new feature isn't implemented.

felipecrs commented:
Superseded by https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/770. Thanks a lot, @stuartrowe !
